### PR TITLE
Wing rotation controller -> state

### PR DESCRIPTION
**Pull request by dewagter**
- same code for 3B and 25
 - removed wing_rotation_controller_servo

### CODE REVIEW

File: conf/modules/wing_rotation_controller_can.xml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE module SYSTEM "module.dtd">
+<module name="wing_rotation_controller_can" dir="rot_wing_drone">
+  <doc>
+    <description>Module to control wing rotation servo command based on prefered angle setpoint</description>
+  </doc>
+  <header>
+    <file name="wing_rotation_controller_can.h"/>
+  </header>
+    <periodic fun="wing_rotation_event()"/>
```
Comment (gautierhattenberger) [inline conf/modules/wing_rotation_controller_can.xml:9]: maybe not the most appropriate name (event/periodic)

File: conf/modules/wing_rotation_controller_can.xml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE module SYSTEM "module.dtd">
+<module name="wing_rotation_controller_can" dir="rot_wing_drone">
+  <doc>
+    <description>Module to control wing rotation servo command based on prefered angle setpoint</description>
+  </doc>
+  <header>
```
Comment (gautierhattenberger) [inline conf/modules/wing_rotation_controller_can.xml:6]: add dependency on can modules ?

File: sw/airborne/modules/rot_wing_drone/wing_rotation_controller_can.c [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+#include "modules/radio_control/radio_control.h"
+#include "firmwares/rotorcraft/guidance/guidance_h.h"
+#include "generated/airframe.h"
+#include "modules/core/abi.h"
+
+#include "state.h"
+
+#include <stdlib.h>
+
+float pprz_angle_step = 9600. / 45.; // CMD per degree
```
Comment (gautierhattenberger) [inline sw/airborne/modules/rot_wing_drone/wing_rotation_controller_can.c:37]: is this a real constant or should it be configurable ?

**Review by gautierhattenberger — CHANGES_REQUESTED** [review]

File: sw/airborne/modules/rot_wing_drone/wing_rotation_adc_sensor.c [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+#include "generated/airframe.h"
+#include "modules/core/abi.h"
+
+#include <stdlib.h>
+#include "mcu_periph/adc.h"
+
+
+
+
+#ifndef ADC_CHANNEL_WING_ROTATION_CONTROLLER_POSITION
```
[the remark this answers is not in the document]
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/wing_rotation_adc_sensor.c:36]: ok

File: sw/airborne/modules/rot_wing_drone/wing_rotation_adc_sensor.c [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+  // SEND ABI Message to ctr_eff_sched and other modules that want Actuator position feedback
+  struct act_feedback_t feedback;
+  feedback.idx =  SERVO_ROTATION_MECH;
+  feedback.position = 0.5 * M_PI - RadOfDeg(wing_angle_deg);
+  feedback.set.position = true;
+
+  // Send ABI message
+  AbiSendMsgACT_FEEDBACK(ACT_FEEDBACK_UAVCAN_ID, &feedback, 1);
+}
```
[the remark this answers is not in the document]
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/wing_rotation_adc_sensor.c:67]: ok

File: sw/airborne/modules/rot_wing_drone/wing_rotation_adc_sensor.c [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+
+  // SEND ABI Message to ctr_eff_sched and other modules that want Actuator position feedback
+  struct act_feedback_t feedback;
+  feedback.idx =  SERVO_ROTATION_MECH;
+  feedback.position = 0.5 * M_PI - RadOfDeg(wing_angle_deg);
+  feedback.set.position = true;
+
+  // Send ABI message
+  AbiSendMsgACT_FEEDBACK(ACT_FEEDBACK_UAVCAN_ID, &feedback, 1);
```
[the remark this answers is not in the document]
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/wing_rotation_adc_sensor.c:66]: Module will be removed as soon as 3B is converted. Will not fix this.

File: conf/modules/wing_rotation_adc_sensor.xml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE module SYSTEM "module.dtd">
+<module name="wing_rotation_adc_sensor" dir="rot_wing_drone">
+  <doc>
+    <description>Module to read wing skew angle from adc</description>
+  </doc>
```
[the remark this answers is not in the document]
Comment (dewagter) [reply conf/modules/wing_rotation_adc_sensor.xml:5]: added docs for the only configure

File: sw/airborne/modules/rot_wing_drone/wing_rotation_adc_sensor.c [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+/** @file "modules/rot_wing_drone/wing_rotation_adc_sensor.c"
+ * @author Dennis van Wijngaarden <D.C.vanWijngaarden@tudelft.nl>
+ * Module to read skew angle from adc sensor
```
[the remark this answers is not in the document]
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/wing_rotation_adc_sensor.c:23]: Will be removed as soon as 3B is converted to CAN wing control.

File: sw/airborne/modules/rot_wing_drone/wing_rotation_adc_sensor.c [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+void wing_rotation_adc_to_deg(void)
+{
+  float adc_wing_rotation = buf_wing_rot_pos.sum / buf_wing_rot_pos.av_nb_sample;
+  float wing_angle_deg = 0.00247111 * adc_wing_rotation - 25.635294;
+
+
+  // SEND ABI Message to ctr_eff_sched and other modules that want Actuator position feedback
+  struct act_feedback_t feedback;
+  feedback.idx =  SERVO_ROTATION_MECH;
```
[the remark this answers is not in the document]
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/wing_rotation_adc_sensor.c:61]: OK

File: sw/airborne/modules/rot_wing_drone/rotwing_state.c [hunk 10]
```diff
@@ -524,6 +552,49 @@ void rotwing_state_skewer(void)
...
+  rotwing_state_skewing.servo_pprz_cmd = servo_pprz_cmd;
+  #endif
+
+  #if USE_NPS
+    actuators_pprz[INDI_NUM_ACT] = (rotwing_state_skewing.servo_pprz_cmd + MAX_PPRZ) / 2.; // Scale to simulation command
+    rotwing_state_skewing.wing_angle_deg = (float) rotwing_state_skewing.servo_pprz_cmd / MAX_PPRZ * 45. + 45.;
+
+    // SEND ABI Message to ctr_eff_sched and other modules that want Actuator position feedback
+    struct act_feedback_t feedback;
+    feedback.idx =  SERVO_ROTATION_MECH;
```
[the remark this answers is not in the document]
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/rotwing_state.c:589]: changed!

File: conf/modules/rotwing_state.xml [hunk 11]
```diff
@@ -18,6 +18,7 @@
   </header>
```
[the remark this answers is not in the document]
Comment (dewagter) [reply conf/modules/rotwing_state.xml:18]: ok

File: sw/airborne/modules/rot_wing_drone/rotwing_state.c [hunk 12]
```diff
@@ -84,6 +84,21 @@
 #define ROTWING_HOV_MOT_OFF_COUNTER 10
 #endif
 
+#ifndef ROTWING_STATE_PD_WING_ROTATION
+#define ROTWING_STATE_PD_WING_ROTATION FALSE
+#endif
+
+#if ROTWING_STATE_PD_WING_ROTATION
```
[the remark this answers is not in the document]
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/rotwing_state.c:91]: Made it static local.

[the remark this answers is not in the document]
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/rotwing_state.c:91]: Renamed the PD_WING_ROTATION to USE_ROTATION_REF_MODEL ... this is not a PD controller but a reference generator.

File: sw/airborne/modules/rot_wing_drone/rotwing_state.c [hunk 13]
```diff
@@ -524,6 +552,49 @@ void rotwing_state_skewer(void)
...
+  // Rotate with second order filter
+  // Smooth accerelation and rate limited setpoint
+  
+  float error_cmd = servo_pprz_cmd - rotwing_state_skew_p_cmd;
+  float speed_sp  = ROTWING_STATE_WING_ROTATION_P_GAIN * error_cmd;
+  float speed_error = speed_sp - rotwing_state_skew_d_cmd;
+  rotwing_state_skew_d_cmd += ROTWING_STATE_WING_ROTATION_D_GAIN * speed_error;
+  rotwing_state_skew_p_cmd += rotwing_state_skew_d_cmd;
+  Bound(rotwing_state_skew_p_cmd, -MAX_PPRZ, MAX_PPRZ);
+  rotwing_state_skewing.servo_pprz_cmd = rotwing_state_skew_p_cmd;
```
[the remark this answers is not in the document]
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/rotwing_state.c:576]: renamed

File: sw/airborne/modules/rot_wing_drone/rotwing_state.c [hunk 14]
```diff
@@ -524,6 +552,49 @@ void rotwing_state_skewer(void)
...
+
+  #else
+  // Directly controlling the wing rotation
+  rotwing_state_skewing.servo_pprz_cmd = servo_pprz_cmd;
+  #endif
+
+  #if USE_NPS
+    actuators_pprz[INDI_NUM_ACT] = (rotwing_state_skewing.servo_pprz_cmd + MAX_PPRZ) / 2.; // Scale to simulation command
```
[the remark this answers is not in the document]
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/rotwing_state.c]: solved

File: sw/airborne/modules/ctrl/eff_scheduling_rot_wing.c [hunk 15]
```diff
@@ -33,7 +33,7 @@
 #include "modules/actuators/actuators.h"
 #include "modules/core/abi.h"
 
-#ifndef SERVO_ROTATION_MECH
+#ifndef SERVO_ROTATION_MECH_IDX
 #error ctrl_eff_sched_rot_wing requires a servo named ROTATION_MECH
```
Comment (dewagter) [inline sw/airborne/modules/ctrl/eff_scheduling_rot_wing.c:37]:
```suggestion
#error ctrl_eff_sched_rot_wing requires a servo named ROTATION_MECH_IDX
```

**Pull request merged**